### ISSUE TITLE
Improvement Tips

### ISSUE THREAD

**Issue by Andrew-Chen-Wang** [shawarmaKoders/Hedwig#9]
Hey @Vibhu-Agarwal 

Saw this project from your report in aioredis. Just a tip for multi-room chats and websockets in general:

- Make it a one socket per user setup. A single user is only connected to one websocket at a time. Think about WhatsApp. How can an app update in real time messages from several chat rooms? Every socket you create creates a new file on your system. Too many users? Not good. Holding open too many sockets? Not good. This means, create a websocket connection via one link: `/{user_id}/`
- If you want to communicate with a user about a real-time update, you can do that through a websocket, specifically a one socket per user setup.

This relies heavily on a specific feature of Redis/aioredis. I challenge you to make it happen ;)

Good luck,
Andrew
Velnota.com

**Comment by Vibhu-Agarwal** [comment]
@Andrew-Chen-Wang Thanks a ton for your improvement tips!

We did consider one websocket per user setup, but it didn't fit our use case.
What we identified is that in our work-flow, **only one chat-room will be active per device at a particular time**.

In Whatsapp, multiple chat rooms are active at all times. The home-screen needs to be updated with new messages coming from different contacts and groups. As you pointed out, opening loads of websocket connections (for each chat-room) simultaneously will use a lot of resources.
The thing is that our app is _not a chatting-app per se_. Messaging features are used occasionally and for a specific purpose. Think of it like a chat-support on Amazon. So what we're doing is that we're opening a websocket connection when one opens a particular chat room and close that connection as soon as one chat-window is exited.

Pros:
\- We don't have to maintain even a single websocket-connection for most of the time a user is using the app.

Actually, the features in Hedwig were originally developed in another project, which is acting as a back-end for an android app but later we pulled out all the messaging part in a separate repository and we've now deployed it separately as well.

👉 The important thing now is that I properly mention the use-case for Hedwig in the README so that it isn't ambiguous.

On second thought, I'm thinking to try out my hands on a (new and separate) project which _will_ be a chatting app per se and then I'll probably get to use a lot of other features of Redis/aioredis and will learn more stuff about designing websocket-connection flows.
